core/legacy/regular-update.js

```diff
@@ -1,93 +1,62 @@
-// interval: number in milliseconds.
-// cb: a callback function that takes an error and data returned by the scraper.
-//
-// To use this from a service:
-//
-// import { promisify } from 'util'
-// import { regularUpdate } from '../../core/legacy/regular-update.js'
-//
-// function getThing() {
-//   return promisify(regularUpdate)({
-//     url: ...,
-//     ...
-//   })
-// }
-//
-// in handle():
-//
-// const thing = await getThing()
-
-function regularUpdate(
-  {
-    url,
-    intervalMillis,
-    json = true,
-    scraper = buffer => buffer,
-    options = {},
-    request = requestModule,
-  },
-  cb
-) {
+/**
+ * Make a HTTP request using an in-memory cache
+ *
+ * @param {object} attrs Refer to individual attrs
+ * @param {string} attrs.url URL to request
+ * @param {number} attrs.intervalMillis Number of milliseconds to keep cached value for
+ * @param {boolean} [attrs.json=true] True if we expect to parse the response as JSON
+ * @param {Function} [attrs.scraper=buffer => buffer] Function to extract value from the response
+ * @param {object} [attrs.options={}] Options to pass to got
+ * @param {Function} [attrs.requestFetcher=fetcher] Custom fetch function
+ * @returns {*} Parsed response
+ */
+async function regularUpdate({
+  url,
+  intervalMillis,
+  json = true,
+  scraper = buffer => buffer,
+  options = {},
+  requestFetcher = fetcher,
+}) {
   const timestamp = Date.now()
   const cached = regularUpdateCache[url]
   if (cached != null && timestamp - cached.timestamp < intervalMillis) {
-    cb(null, cached.data)
-    return
+    return cached.data
   }
-  request(url, options, (err, res, buffer) => {
-    if (err != null) {
-      cb(
-        new Inaccessible({
-          prettyMessage: 'intermediate resource inaccessible',
-          underlyingError: err,
-        })
-      )
-      return
-    }
 
-    if (res.statusCode < 200 || res.statusCode >= 300) {
-      cb(
-        new InvalidResponse({
-          prettyMessage: 'intermediate resource inaccessible',
-        })
-      )
-    }
+  const { buffer } = await checkErrorResponse({})(
+    await requestFetcher(url, options)
+  )
 
-    let reqData
-    if (json) {
-      try {
-        reqData = JSON.parse(buffer)
-      } catch (e) {
-        cb(
-          new InvalidResponse({
-            prettyMessage: 'unparseable intermediate json response',
-            underlyingError: e,
-          })
-        )
-        return
-      }
-    } else {
-      reqData = buffer
-    }
-
-    let data
+  let reqData
+  if (json) {
     try {
-      data = scraper(reqData)
+      reqData = JSON.parse(buffer)
     } catch (e) {
-      cb(e)
-      return
+      throw new InvalidResponse({
+        prettyMessage: 'unparseable intermediate json response',
+        underlyingError: e,
+      })
     }
+  } else {
+    reqData = buffer
+  }
 
-    regularUpdateCache[url] = { timestamp, data }
-    cb(null, data)
-  })
+  const data = scraper(reqData)
```

Comment:
it's a small diff but admittedly my eyes are having difficulty following it, though curious if there's any potential behavioral concerns since the `scraper` call site is no longer happening within a `try` block?

Reply:
Yeah this took me a while to unpick too.

You'll notice that the old `regularUpdate` function takes a callback function as a param (`cb`) in the sig, but when we're invoking `regularUpdate()` we are not actually passing it a callback. This is because we're wrapping it in [`promisify()`](https://github.com/nodejs/node/blob/657fb9a77ca36f729da484f55899dad7a13759b0/lib/internal/util.js#L317-L365) from the standard library. One of the things this does is it passes our function a callback that [rejects the promise if we pass it an error](https://github.com/nodejs/node/blob/657fb9a77ca36f729da484f55899dad7a13759b0/lib/internal/util.js#L341-L343)

So

```js
function regularUpdate(params, cb) {
  try {
    // do a thing
  } catch (e) {
    cb(e)
    return
  }
}
```

is there to catch any exception we throw and pass it into the callback function, but its not actually handling it pre-se - its just allowing us to catch it when we resolve the promise (maybe "bubbling" it is the right word for this?). i.e: it allows us to do:

```js
try:
  await promisify(regularUpdate)(params)
} catch (e) {
  // either handle the error cleanly, or ignore it and leave it sentry
}
```

Now that `regularUpdate` is an async function, we can safely just throw in an async function, so

```js
function regularUpdate(params, cb) {
  try {
    // do a thing
  } catch (e) {
    cb(e)
    return
  }
}
```

in callback-land can just be replaced with

```js
async function regularUpdate(params) {
  // do a thing
}
```

in async/await-land because:

```js
try:
  await regularUpdate(params)
} catch (e) {
  // either handle the error cleanly, or ignore it and leave it sentry
}
```

is already a thing.

Reply:
got it, thanks for the explanation!